### PR TITLE
Implement spans for metrics profiling

### PR DESCRIPTION
**Pull request by roby2014**
# Description

See linked issue for proposed implementation, but in short:
- Create new namespace `tel` with metrics, tracing and logging.
- Create a global "telemetry" level for logs and spans.
- Create span utilities to keep track of scopes and execution times (for now).
- Associate metrics and logs to spans.

## Checklist

- [x] Self-review changes.
- [x] Evaluate impact on the documentation.
- [x] Ensure test coverage.
- [x] Write new samples.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -30,6 +32,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Spot shadows cause light range cutoff (#1312, **@tomas7770**).
 
 
+### Changed
+
+- Move logger to telemetry namespace (#1265, **@roby2014**).
+- Move metrics to telemetry namespace (#1265, **@roby2014**).
```
Comment (RiscadoA) [inline CHANGELOG.md:38]: This should be moved to the previous ### Changes section, before the ### Fixed section.
Also, I think you can merge those two, specially considering they come from the same issue.

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Generic Camera component to hold projection matrix (#1331, **@mkuritsu**).
 - Initial application debugging through Tesseratos (#1303, **@RiscadoA**).
 - Print stacktrace with *cpptrace* on calls to CUBOS_FAIL (#1172, **@RiscadoA**).
+- Create global telemetry level (for tracing/logging) (#1265, **@roby2014**).
```
Comment (RiscadoA) [inline CHANGELOG.md:22]:
```suggestion
- Global telemetry level (for tracing/logging) (#1265, **@roby2014**).
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -19,6 +19,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Implemented tracing (#1265, **@roby2014**).
```
Comment (RiscadoA) [inline CHANGELOG.md:23]:
```suggestion
- Spans for tracing and profiling (#1265, **@roby2014**).
```

File: core/include/cubos/core/tel/level.hpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
+
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/level.hpp:1]:
```suggestion
```

File: core/include/cubos/core/tel/metrics.hpp [hunk 5]
```diff
@@ -96,7 +96,7 @@ namespace cubos::core
 /// @note The macro uses `__FUNCTION__` to register the metric name.
 
 #ifdef CUBOS_PROFILING
-#define CUBOS_PROFILE() ::cubos::core::ScopeProfiler _profilingTimeTracker(__FUNCTION__, __FILE__, __LINE__)
+#define CUBOS_PROFILE() ::cubos::core::tel::ScopeProfiler _profilingTimeTracker(__FUNCTION__, __FILE__, __LINE__)
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/metrics.hpp:99]: This is kind of useless now, no? Considering we now have spans
Should we remove it? We should only add the `begin` and `end` metrics to spans if `CUBOS_PROFILING` is enabled.

File: core/include/cubos/core/tel/tracing.hpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
+/// @file
+/// @brief Tracing macros.
+/// @ingroup core-tel
+
+#pragma once
+
+#include <chrono>
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:7]: You're not using this here, so you can move it to the .cpp file

File: core/include/cubos/core/tel/tracing.hpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+#include <string>
+
+#include <cubos/core/tel/level.hpp>
+
+#define SPAN(a, b) SPAN_INNER(a, b)
+#define SPAN_INNER(a, b) a##b
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:13]: You don't need these macros right?

File: core/include/cubos/core/tel/tracing.hpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+namespace cubos::core::tel
+{
+    /// @brief Represents the unique identifier and path for a span.
+    struct SpanId
+    {
+        std::size_t id;   ///< Unique identifier.
+        std::string path; ///< Span path.
+    };
+
+    /// @brief A guard object that ensures a span ends when it goes out of scope.
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:41]:
```suggestion
    /// @brief A guard object that automatically ends the current span when it goes out of scope.
```

File: core/include/cubos/core/tel/tracing.hpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+#define CUBOS_SPAN_TRACE(name) cubos::core::tel::SpanManager::begin(name, cubos::core::tel::Level::Trace)
+
+/// @}
+
+namespace cubos::core::tel
+{
+    /// @brief Represents the unique identifier and path for a span.
+    struct SpanId
+    {
+        std::size_t id;   ///< Unique identifier.
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:37]:
```suggestion
        std::size_t id;   ///< Identifier which distinguishes between instances of spans with the same path.
```
It's not really unique - two threads will produce spans with the same ids. Both fields are needed to get uniqueness.

File: core/include/cubos/core/tel/tracing.hpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+/// @brief Constructs a new trace span.
+/// @param name Span name.
+#define CUBOS_SPAN_TRACE(name) cubos::core::tel::SpanManager::begin(name, cubos::core::tel::Level::Trace)
+
+/// @}
+
+namespace cubos::core::tel
+{
+    /// @brief Represents the unique identifier and path for a span.
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:34]:
```suggestion
    /// @brief Uniquely identifies a span.
```

File: core/include/cubos/core/tel/tracing.hpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        ///
+        /// If @ref CUBOS_SPAN_LEVEL is higher than the set level, entries below it are still not
+        /// registered.
+        ///
+        /// @param level New span level.
+        static void level(Level level);
+
+    private:
+        // Private constructor.
+        SpanManager() = default;
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:72]:
```suggestion
```
You end up not instantiating objects of this class, so you can remove this

File: core/include/cubos/core/tel/tracing.hpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        /// @return Span* Pointer to the current active span.
+        static SpanId current();
+
+        /// @brief Sets the tracing level.
+        ///
+        /// If @ref CUBOS_SPAN_LEVEL is higher than the set level, entries below it are still not
+        /// registered.
+        ///
+        /// @param level New span level.
+        static void level(Level level);
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:68]:
```suggestion
```
Redundant now, as we already have the global level function.

File: core/include/cubos/core/tel/tracing.hpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        ~SpanGuard();
+    };
+
+    /// @brief Manages the creation and ending of spans.
+    class SpanManager
+    {
+    public:
+        /// @brief Begins a new span.
+        /// @param s Pointer to the span to be entered.
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:52]:
```suggestion
        /// @param name Span name.
        /// @param level Span level.
```

File: core/samples/tel/metrics/page.md [hunk 14]
```diff
@@ -1,15 +1,15 @@
-# Metrics {#examples-core-metrics}
+# Metrics {#examples-core-tel-metrics}
 
 @brief Using the metrics system.
 
 @see Full source code [here](https://github.com/GameDevTecnico/cubos/tree/main/core/samples/metrics).
 
-Registering metrics and profiling in **Cubos** can be done through the metric macros defined in @ref core/metrics.hpp and the @ref core::Metrics class.
+Registering metrics and profiling in **CUBOS.** can be done through the metric macros defined in @ref core/tel/metrics.hpp and the @ref core::tel::Metrics class.
```
Comment (RiscadoA) [inline core/samples/tel/metrics/page.md:7]:
```suggestion
Registering metrics and profiling in **Cubos** can be done through the metric macros defined in @ref core/tel/metrics.hpp and the @ref core::tel::Metrics class.
```

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thank you for working on this! Seems mostly good to me, just some nitpicks.
Also, you need to add a module.dox for the new `core-tel` module. Check out other modules (e.g cubos/core/gl/module.dox)

File: core/include/cubos/core/tel/tracing.hpp [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        /// @brief Begins a new span.
+        /// @param name Span name.
+        /// @param level Span level.
+        static void begin(const std::string& name, Level level);
+
+        /// @brief Ends the current active span.
+        static void end();
+
+        /// @brief Gets the current active span.
+        /// @return Span* Pointer to the current active span.
```
Comment (RiscadoA) [inline core/include/cubos/core/tel/tracing.hpp:69]:
```suggestion
        /// @return Identifier of the current span.
```

**Review by RiscadoA — CHANGES_REQUESTED** [review]

**Review by diogomsmiranda — APPROVED** [review]
I don't see much code of this part of CUBOS so can't be of much help, but looks like most of files changed were just refactor of the code to fit the new namespace. Found nothing very wrong when looking at the implementation and docs.
If all the other requested changes were already addressed then LGTM!

Very useful feature aswell!!

**Review by RiscadoA — APPROVED** [review]
LGTM! Ty for the crunch :pray:

**Review by Dageus — APPROVED** [review]

**Pull request merged**